controller/lib/core/nvparam.cpp

```diff
@@ -0,0 +1,218 @@
+    if (!nvparam_addr) {
+      nvparam_addr = addr;
+      continue;
+    }
+
+    // I've found multiple valid blocks.  This normally shouldn't
+    // happen but could if power is lost at just the right time
+    // when updating flash.
+    //
+    // I'll check the counter to see which was is newer and keep
+    // that address.  I'll also invalidate the older one.
+    NVparams *a = reinterpret_cast<NVparams *>(nvparam_addr);
+    NVparams *b = reinterpret_cast<NVparams *>(addr);
+
+    int8_t diff = static_cast<int8_t>(b->count - a->count);
+    if (diff > 0) {
+      Invalidate(nvparam_addr);
+      nvparam_addr = addr;
+    } else
+      Invalidate(addr);
+  }
+
+  // If I found a valid parameter area, I'm done
+  if (nvparam_addr)
+    return;
+
+  // If no valid storage was found then I'll just create one with
+  // zero values.
+  NVparams blank;
+  memset(&blank, 0, sizeof(blank));
+  blank.mark = good_mark;
+  blank.crc = BlockCRC(&blank);
+
+  Hal.FlashErasePage(flash_params_start);
+  Hal.FlashWrite(flash_params_start, reinterpret_cast<uint32_t *>(&blank),
+                 sizeof(blank) / sizeof(uint32_t));
+  nvparam_addr = flash_params_start;
+}
+
+const NVparams *FindStore(void) {
+  return reinterpret_cast<const NVparams *>(nvparam_addr);
+}
+
+bool NVparamsUpdtOff(uint32_t offset, const void *value, uint8_t len) {
+
+  // Make sure the passed pointer is pointing to somewhere
+  // in the current page and isn't in the reserved first 8 bytes
+  if ((offset < 8) || (offset >= nvparam_size))
+    return false;
+
+  // Copy the current parameter block to RAM
+  NVparams temp;
+  memcpy(&temp, reinterpret_cast<void *>(nvparam_addr), nvparam_size);
+
+  // Update the contents in RAM
+  memcpy(reinterpret_cast<uint8_t *>(&temp) + offset, value, len);
+
+  // Make sure the reserved area is kept at zero
+  memset(&temp.rsvd, 0, sizeof(temp.rsvd));
+
+  temp.count++;
+  temp.mark = good_mark;
+  temp.crc = BlockCRC(&temp);
+
+  uint32_t addr = nvparam_addr + nvparam_size;
+
+  bool ok = SaveBlock(&temp, addr);
+
+  // That shouldn't really fail, but if it does for some reason
+  // try storing in the first location of the other flash page.
+  // That will cause an erase first, so is more likely to succeed
+  if (!ok) {
+    addr = flash_params_start;
+    if (nvparam_addr < flash_params_start + flash_page_size)
+      addr += flash_page_size;
+    ok = SaveBlock(&temp, addr);
+  }
+
+  if (!ok)
+    return false;
+
+  // After successfully storing and validating the new parameters,
+  // Invalidate the old ones
+  Invalidate(nvparam_addr);
+  nvparam_addr = addr;
+
+  return true;
+}
+
+static bool SaveBlock(NVparams *blk, uint32_t addr) {
+  // If the address is the start of a page, first erase the page
+  if ((addr & (flash_page_size - 1)) == 0) {
+    bool ok = Hal.FlashErasePage(addr);
+    if (!ok)
+      return false;
+  }
+
+  int ok =
+      Hal.FlashWrite(addr, (uint32_t *)blk, nvparam_size / sizeof(uint32_t));
```

Comment:
This is a strict-aliasing violation, which can and does cause undefined behavior (https://stackoverflow.com/questions/2958633/gcc-strict-aliasing-and-horror-stories).  Especially because we do link-time optimization, the compiler can observe this and make the wrong aliasing assumption.

Specifically the violation is to read/write from this pointer as an NVparams* and a uint32_t*.  That is, the violation is not actually the cast, but the read inside of FlashWrite.

Anywho, FlashWrite should probably take a char* or uint8_t*, that fixes it.

Reply:
I don't really understand the problem.  I prefer the uint32_t * on the call to FlashWrite because that allows me to access the data as 32-bit ints directly.  When I write to the flash I need to write a pair of 32-bit ints at a time.
Is it because I accidentally let a C style cast slip through?  Does changing this to reinterpret_cast resolver the issue?

Reply:
> I don't really understand the problem

LMK if this helps?  https://gist.github.com/shafik/848ae25ee209f698763cffee272a58f8

> Does changing this to reinterpret_cast resolver the issue?

Unfortunately no.

In very brief, if you access a memory location through a pointer of type T*, you cannot access it through a pointer of type U* unless either (a) there is an intervening deallocation at that address, (b) either T or U is char or unsigned char, or (c) the access is through memcpy or memzero.

This is true in both C and C++.

Reply:
OK, I've changed FlashWrite to use a void* which should be safe.